cmd/gpu_plugin/rm/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,417 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rm
+
+import (
+	"context"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	dpapi "github.com/intel/intel-device-plugins-for-kubernetes/pkg/deviceplugin"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+	podresourcesv1 "k8s.io/kubelet/pkg/apis/podresources/v1"
+	"k8s.io/kubernetes/pkg/kubelet/apis/podresources"
+)
+
+const (
+	deviceType       = "i915"
+	fullResourceName = "gpu.intel.com/" + deviceType
+	monitorID        = "all"
+
+	gasTSAnnotation   = "gas-ts"
+	gasCardAnnotation = "gas-container-cards"
+
+	grpcAddress    = "unix:///var/lib/kubelet/pod-resources/kubelet.sock"
+	grpcBufferSize = 4 * 1024 * 1024
+	grpcTimeout    = 5 * time.Second
+
+	retryTimeout = 1 * time.Second
+)
+
+// Errors.
+var (
+	errZeroPending = zeroPendingErr{}
+	errRetry       = retryErr{}
+)
+
+type retryErr struct{}
+
+func (e retryErr) Error() string {
+	return "things didn't work out, but perhaps a retry will help"
+}
+
+type zeroPendingErr struct{}
+
+func (e zeroPendingErr) Error() string {
+	return "there are no pending pods anymore in this node"
+}
+
+type podCandidate struct {
+	pod                 *v1.Pod
+	name                string
+	allocatedContainers int
+	allocationTargetNum int
+}
+
+type DeviceInfoSubset struct {
+	nodes  []pluginapi.DeviceSpec
+	mounts []pluginapi.Mount
+}
+
+type ResourceManager struct {
+	mutex       sync.RWMutex // for devTree updates during scan
+	deviceInfos map[string]DeviceInfoSubset
+	nodeName    string
+	clientset   *kubernetes.Clientset
+}
+
+func NewResourceManager() *ResourceManager {
+	clientset := getClientset()
+
+	if clientset == nil {
+		klog.Warning("no clientset available, resource manager not created")
+		return nil
+	}
+
+	rm := ResourceManager{
+		nodeName:  os.Getenv("NODE_NAME"),
+		clientset: clientset,
+	}
+
+	klog.Info("GPU device plugin resource manager enabled")
+
+	return &rm
+}
+
+// ReallocateWithFractionalResources runs the fractional resource logic.
+// This intentionally only logs errors and returns with the default device allocation, in case any errors are hit. This is to avoid
+// clusters filling up with unexpected admission errors.
+func (rm *ResourceManager) ReallocateWithFractionalResources(request *pluginapi.AllocateRequest, allocateResponse *pluginapi.AllocateResponse) {
+	var err error
+	if !isInputOk(request, allocateResponse) {
+		// it is better to leave allocated gpu devices as is and return
+		return
+	}
+
+	podCandidate, err := rm.findAllocationPodCandidate()
+	if errors.Is(err, errRetry) {
```

Comment:
```suggestion
	if errors.Is(err, retryErr) {
```

Reply:
Again the naming comes from one linter rule which suggests errors to be named with prefix "err". However since they are not going to be globals, and since that linter rule is not enabled in this project, making it retryErr also is fine.